src/lidar_tools/dsm_functions.py

```diff
@@ -223,106 +339,342 @@ def create_dem_stage(dem_filename='dem_output.tif', pointcloud_resolution=1.,
+    dem_diff = lidar_da_masked - cop30_da_masked
+    ## Mask out bare and sparse vegetation class
+    bare_sparse_mask = worldcover_da_masked == 60
+    dem_diff_bare = dem_diff.where(bare_sparse_mask,np.nan)
+    median_diff = np.nanmedian(dem_diff_bare.values)
+    print(f"Observed difference between COP30 EGM2008 and 3DEP LiDAR DSM over bareground and sparse vegetation surfaces is {median_diff:.2f} m")
+    if np.abs(median_diff) <= bare_diff_tolerance:
+        #this means that both COP30 and 3DEP LiDAR DSM are with respect to geoid
+        print("Looks like the 3DEP height estimates are with respect to geiod, will apply vertical datum shift to return heights with respect to ellipsoid")
+        out = True
+    else:
+        #this means that 3DEP LiDAR DSM is with respect to ellipsoid
+        print("Looks like the 3DEP height estimates are already with respect to ellipsoid, geoid to ellipsoid transformation will not be attempted")
+        out = False
+    return out
 
+def check_raster_validity(raster_fn: str) -> bool:
+    """
+    Check if a raster file is valid and can be opened using rioxarray and CRS check
 
+    Parameters
+    ----------
+    raster_fn : str
+        Path to the raster file.
 
+    Returns
+    -------
+    bool
+        True if the raster file is valid, False otherwise.
+    """
+    da = rioxarray.open_rasterio(raster_fn,masked=True).squeeze()
+    if da.rio.crs is None:
+        #print(f"Raster {raster_fn} does not have a valid CRS.")
+        out = False
+    else:
+        #print(f"Raster {raster_fn} has a valid CRS.")
+        out = True
+    return out   
+def gdal_warp(src_fn: str,
+                dst_fn: str, 
+                src_srs: str, 
+                dst_srs: str, 
+                res: float = 1.0,
+                resampling_alogrithm: str ='cubic') -> None:
+    tolerance = 0
+    resampling_mapping = {"nearest":  gdalconst.GRA_NearestNeighbour, "bilinear": gdalconst.GRA_Bilinear,
+                  "cubic": gdalconst.GRA_Cubic, "cubic_spline": gdalconst.GRA_CubicSpline}
+    resampling_alg = resampling_mapping[resampling_alogrithm]
+    ds = gdal.Warp(dst_fn, src_fn,
+                   resampleAlg=resampling_alg,
+                   srcSRS=src_srs, xRes=res, yRes=res,
+                   dstSRS=dst_srs, errorThreshold=tolerance,
+                   targetAlignedPixels=True,
+                   creationOptions=['COMPRESS=LZW','TILED=YES','COPY_SRC_OVERVIEWS=YES'],
+                   callback=gdal.TermProgress_nocb)
+    ds = None
+def gdal_add_overview(raster_fn: str) -> None:
+    """
+    Add overviews to a raster file using GDAL.
+
+    Parameters
+    ----------
+    raster_fn : str
+        Path to the raster file.
+    """
+    ds = gdal.Open(raster_fn, 1)
+    gdal.SetConfigOption('COMPRESS_OVERVIEW', 'DEFLATE')
+    ds.BuildOverviews("GAUSS",[2,4,8,16,32,64,128,256,512,1024,2048,4096],
```

Comment:
That's a lot of overviews! Is it necessary to go beyond 64?

Reply:
Okies, updated here: [4416695](https://github.com/uw-cryo/lidar_tools/commit/44166959be6017d7f00d185f981b52552e123c3d)